src/sagemaker/workflow/airflow.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import os
+
+from sagemaker.estimator import Framework
+from sagemaker.amazon.amazon_estimator import AmazonAlgorithmEstimatorBase, RecordSet
+from sagemaker.job import _Job
+from sagemaker.utils import base_name_from_image, airflow_name_from_base
+from sagemaker.model import DIR_PARAM_NAME, SCRIPT_PARAM_NAME, CLOUDWATCH_METRICS_PARAM_NAME, \
+    CONTAINER_LOG_LEVEL_PARAM_NAME, JOB_NAME_PARAM_NAME, SAGEMAKER_REGION_PARAM_NAME
+
+
+def prepare_framework(estimator, s3_operations, default_bucket):
+    """
+    Prepare S3 operations (specify where to upload source_dir) and environment variables
+        related to framework.
+
+    Args:
+        estimator: The framework estimator to get information from and update.
+        s3_operations: The dict to specify s3 operations (upload source_dir).
+        default_bucket: The default bucket to use in training.
+    """
+    bucket = estimator.code_location if estimator.code_location else default_bucket
+    key = '{}/source/sourcedir.tar.gz'.format(estimator._current_job_name)
+    script = os.path.basename(estimator.entry_point)
+    if estimator.source_dir and estimator.source_dir.lower().startswith('s3://'):
+        code_dir = estimator.source_dir
+    else:
+        code_dir = 's3://{}/{}'.format(bucket, key)
+        s3_operations['S3Upload'] = [{
+            'Path': estimator.source_dir or script,
+            'Bucket': bucket,
+            'Key': key,
+            'Tar': True
+        }]
+    estimator._hyperparameters[DIR_PARAM_NAME] = code_dir
+    estimator._hyperparameters[SCRIPT_PARAM_NAME] = script
+    estimator._hyperparameters[CLOUDWATCH_METRICS_PARAM_NAME] = estimator.enable_cloudwatch_metrics
+    estimator._hyperparameters[CONTAINER_LOG_LEVEL_PARAM_NAME] = estimator.container_log_level
+    estimator._hyperparameters[JOB_NAME_PARAM_NAME] = estimator._current_job_name
+    estimator._hyperparameters[SAGEMAKER_REGION_PARAM_NAME] = estimator.sagemaker_session.boto_region_name
+
+
+def prepare_amazon_algorithm_estimator(estimator, inputs):
```

Comment:
Inputs should be optional

Reply:
Updated. And remove the exception raised. I found I should not raise exception here since it might be other value.